perfkitbenchmarker/providers/gcp/gce_network.py

```diff
@@ -235,27 +243,72 @@ def __init__(self, network_spec):
+
+    if FLAGS.gce_subnet_region:  # Overrides auto network when set. Creates one VPC network in single region.
+      self.net_type = 'single'
+      self.subnet_addr = FLAGS.gce_subnet_addr
+      self.net_name = network.BaseNetwork.FormatCidrString(FLAGS.gce_subnet_addr)
+      name = FLAGS.gce_network_name or 'pkb-network-%s-%s-%s' % (self.net_type, self.net_name, FLAGS.run_uri)
+    if network_spec.cidr:  # Creates multiple VPC networks when set.
+      self.net_type = 'multi'
+      self.subnet_addr = network_spec.cidr
+      self.net_name = network.BaseNetwork.FormatCidrString(network_spec.cidr)
+      name = FLAGS.gce_network_name or 'pkb-network-%s-%s-%s' % (self.net_type, self.net_name, FLAGS.run_uri)
+
+    self.subnet_region = FLAGS.gce_subnet_region if network_spec.cidr is None else util.GetRegionFromZone(network_spec.zone)
+    mode = 'auto' if self.subnet_region is None else 'custom'
     self.network_resource = GceNetworkResource(name, mode, self.project)
-    if FLAGS.gce_subnet_region is None:
+    if self.subnet_region is None:
       self.subnet_resource = None
     else:
       self.subnet_resource = GceSubnetResource(name, name,
-                                               FLAGS.gce_subnet_region,
-                                               FLAGS.gce_subnet_addr,
+                                               self.subnet_region,
+                                               self.subnet_addr,
                                                self.project)
-    firewall_name = 'default-internal-%s' % FLAGS.run_uri
+
+    self.all_nets = []  # Holds the different networks we learn about.
+    self.external_nets_rules = {}  # Holds FW rules for any external subnets.
+    self.default_subnet = None  # Holds the default subnet for this run if used.
+
+    firewall_name = '%s-%s-%s-%s' % (self.net_type, 'internal', self.net_name, FLAGS.run_uri)
     self.default_firewall_rule = GceFirewallRule(
-        firewall_name, self.project, ALLOW_ALL, name, NETWORK_RANGE)
+        firewall_name, self.project, ALLOW_ALL, name, self.subnet_addr)
+    self.all_nets.append(self.net_name)
+    if self.net_type != 'multi':  # We're using the default network.
+      self.default_subnet = self.subnet_addr
+
+    # Add firewall rules for other networks if they exist.
+    if network_spec.custom_subnets:
+      for net_group, cidr in network_spec.custom_subnets.items():
+        this_net_string = None
+        this_net = None
+        if cidr is None and self.default_subnet is None:  # only add default rule once
+          self.default_subnet = FLAGS.gce_subnet_addr if FLAGS.gce_subnet_region else NETWORK_RANGE
+          this_net_string = network.BaseNetwork.FormatCidrString(self.default_subnet)
+          this_net = self.default_subnet
+        elif cidr:
+          this_net_string = network.BaseNetwork.FormatCidrString(cidr)
+          this_net = cidr
+        if this_net and this_net_string not in self.all_nets:
```

Comment:
is it possible, this_net and this_net_string is still not initialized at this point?

Reply:
Fixed - refactored into separate methods.